File: test/ModularPipelines.UnitTests/NotInParallelTestsWithMultipleConstraintKeys.cs
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests;

[TUnit.Core.NotInParallel]
public class NotInParallelTestsWithMultipleConstraintKeys : TestBase
{
    private static readonly NotInParallelTracker Tracker = new();

    [ModularPipelines.Attributes.NotInParallel("A")]
    public class Module1 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithMultipleConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["Module2"];
    }

    [ModularPipelines.Attributes.NotInParallel("A", "B")]
    public class Module2 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithMultipleConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["Module1", "Module3"];
    }

    [ModularPipelines.Attributes.NotInParallel("B", "C")]
    public class Module3 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithMultipleConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["Module2"];
    }

    [ModularPipelines.Attributes.NotInParallel("D")]
    public class Module4 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithMultipleConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => [];
        protected override int DelayMs => 50;
    }

    [Test]
    public async Task NotInParallel_If_Any_Modules_Executing_With_Any_Of_Same_ConstraintKey()
    {
        Tracker.Reset();

        await TestPipelineHostBuilder.Create()
            .AddModule<Module1>()
            .AddModule<Module2>()
            .AddModule<Module3>()
            .AddModule<Module4>()
            .ExecutePipelineAsync();

        await Assert.That(Tracker.Violations).IsEmpty();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Execution/NotInParallelTestsWithMultipleConstraintKeys.fs
namespace ModularPipelines.UnitTests.FSharp.Execution

open System.Collections.Generic
open ModularPipelines.Extensions
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module private NotInParallelTestsWithMultipleConstraintKeysData =
    let tracker = NotInParallelTracker()

[<ModularPipelines.Attributes.NotInParallel("A")>]
type Module1() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithMultipleConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["Module2"] :> IEnumerable<string>

[<ModularPipelines.Attributes.NotInParallel("A", "B")>]
type Module2() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithMultipleConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["Module1"; "Module3"] :> IEnumerable<string>

[<ModularPipelines.Attributes.NotInParallel("B", "C")>]
type Module3() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithMultipleConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["Module2"] :> IEnumerable<string>

[<ModularPipelines.Attributes.NotInParallel("D")>]
type Module4() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithMultipleConstraintKeysData.tracker
    override _.ConflictingModuleNames = [] :> IEnumerable<string>
    override _.DelayMs = 50

type NotInParallelTestsWithMultipleConstraintKeys() =
    inherit TestBase()

    [<Test>]
    member _.NotInParallel_If_Any_Modules_Executing_With_Any_Of_Same_ConstraintKey() = async {
        NotInParallelTestsWithMultipleConstraintKeysData.tracker.Reset()

        let! _ = TestPipelineHostBuilder.Create()
                    .AddModule<Module1>()
                    .AddModule<Module2>()
                    .AddModule<Module3>()
                    .AddModule<Module4>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
        do! check(Assert.That<string>(NotInParallelTestsWithMultipleConstraintKeysData.tracker.Violations).IsEmpty())
    }
